File: gcf/test/gcf/src/gcf/cross_call/Binary.cpp
#include "Binary.h"
#include <cassert>

namespace cross_call
{
    //////////////////////////////////////////////////////////////////////////
    // NormalReader
    Reader::Reader(serialize::IBinaryReadStream* stream)
        : serialize::BinaryReader(stream)
    {
    }

    bool Reader::StructBegin(int32_t code, const char* name/* = nullptr*/)
    {
        int32_t check = 0;
        return _stream->Read(&check, 4) && check == code;
    }

    bool Reader::StructEnd()
    {
        return true;
    }

    //////////////////////////////////////////////////////////////////////////
    // NormalWriter
    Writer::Writer(serialize::IBinaryWriteStream* stream)
        : serialize::BinaryWriter(stream)
    {
    }

    bool Writer::StructBegin(int32_t code, const char* name/* = nullptr*/)
    {
        return _stream->Write(&code, 4);
    }

    bool Writer::StructEnd()
    {
        return true;
    }
}

File: gcf/test/test_cs/gcf/CrossCall/Binary.cs
/*
 * 二进制序列化
*/

namespace CrossCall
{
    /*
     * 二进制数据Reader 
     */
    public class BinaryReader : Serialize.BinaryReader
    {
        public BinaryReader(System.IO.BinaryReader reader) : base(reader)
        {
        }

        public override bool StructBegin(int code, string name = null)
        {
            int check = 0;
            if (!Read(ref check, null))
                return false;

            return check == code;
        }
    }

    /*
     * 二进制数据Writer
     */
    public class BinaryWriter : Serialize.BinaryWriter
    {
        public BinaryWriter(System.IO.BinaryWriter writer) : base(writer)
        {
        }

        public override bool StructBegin(int code, string name = null)
        {
            return Write(code);
        }

        public override bool StructEnd()
        {
            return true;
        }
    }
}
